Remove IP Option for packet forwarded to the backend

sypakine opened the pull request:
Remove IP Option for packet forwarded to the backend [Optional]

Before forwarding the traffic to the endpoint
1. strip the DSR IP option
2. recalculate the L3 checksum

Fixes: #29900

This change is a fix for backends that drop packets with IP options. 

---

Change to XDP helper is necessary, as tail_lb_ipv4 call hierarchy cannot include calls to abort:

> error: include/bpf/ctx/xdp.h:329:4: in function tail_lb_ipv4 i32 (%struct.xdp_md*): A call to built-in function 'abort' is not supported.
>
> make[1]: *** [Makefile:40: xdp_nodeport_lb4_dsr_lb.o] Error 1

Update tests for backend to not check for IP option. This is implicitly checked by accessing the expected location of the L4 header information. 

---

release-note:

> Minor Changes:
> - Remove IP Option before forwarding the traffic to the backend endpoint [Optional]

asauber commented on bpf/lib/nodeport.h:
```diff
@@ -2081,6 +2081,45 @@ nodeport_extract_dsr_v4(struct __ctx_buff *ctx,
+	iph_old = *(__u32 *)ip4;
+	/*  Remove option */
+	ip4->ihl -= sizeof(opt) >> 2;
+    ip4->tot_len = bpf_htons(bpf_ntohs(ip4->tot_len) - sizeof(opt));
```
indentation is inconsistent here, also on line 2090 above

sypakine replied:
Thanks; done.

asauber commented on bpf/tests/tc_nodeport_lb4_dsr_backend.c:
```diff
@@ -13,6 +13,7 @@
 #define ENABLE_IPV4
 #define ENABLE_NODEPORT
 #define ENABLE_DSR		1
+#define REMOVE_DSR_IP_OPTION 1
```
indentation could be made consistent here as well

sypakine replied:
Thanks, done. LMK if this is what you had in mind.

asauber commented on bpf/tests/tc_nodeport_lb4_dsr_backend.c:
```diff
@@ -181,7 +182,6 @@ int nodeport_dsr_backend_setup(struct __ctx_buff *ctx)
 CHECK("tc", "tc_nodeport_dsr_backend")
 int nodeport_dsr_backend_check(struct __ctx_buff *ctx)
 {
-	struct dsr_opt_v4 *opt;
```
Rather than remove the test logic related to the DSR IP Option, an additional test should be added which tests the logic specifically changed by `#define REMOVE_DSR_IP_OPTION 1`. 

It's still useful to test that the Option isn't mangled when we want it preserved.

sypakine replied:
Thanks; done. Created a new test file (as the `REMOVE_DSR_IP_OPTION` can only be defined once). LMK if this is what you had in mind.

asauber replied:
Thanks, that is what I had in mind

ti-mo reviewed (approved):
OK for @cilium/loader

asauber commented on bpf/tests/tc_nodeport_lb4_dsr_backend_remove_ip_option.c:
```diff
@@ -0,0 +1,411 @@
+#define ENABLE_NODEPORT
+#define ENABLE_DSR				1
+#define REMOVE_DSR_IP_OPTION	1
+#define DSR_ENCAP_GENEVE		3
```
Overall PR looks good. You've introduced some indentation issues here.

ti-mo commented on bpf/include/bpf/ctx/xdp.h:
```diff
@@ -326,7 +326,7 @@ ctx_adjust_hroom(struct xdp_md *ctx, const __s32 len_diff, const __u32 mode,
 				ret = -EFAULT;
 			break;
 		default:
-			__throw_build_bug();
```
You've mentioned in the commit message that this blew up during build, but that seems exactly like this statement's purpose. Doesn't this change invalidate any assumptions made elsewhere in the code? Did you investigate why this fell over?

sypakine replied:
I don't have full context, so thanks for the push-back on this.

After discussussing https://github.com/cilium/cilium/issues/29900#issuecomment-1881146704 with @julianwiedmann, we may be able to avoid adjusting the headroom all together.

If the above works, I will remove this from the change and resolve the comment.

meyskens reviewed (approved):
Changes to agent owned files are LGTM

julianwiedmann commented on bpf/lib/common.h:
```diff
@@ -748,6 +748,16 @@ enum metric_dir {
+ * 
+ * See: https://www.rfc-editor.org/rfc/rfc791.html
+ */
+#define NOP_IPV4_OPT 0x01
```
Note that `bpf/include/linux/ip.h` already gives us `IPOPT_NOOP` (and `IPOPT_END`).

sypakine replied:
Done.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2078,6 +2078,47 @@ nodeport_extract_dsr_v4(struct __ctx_buff *ctx,
+	opt.type = NOP_IPV4_OPT;
+	opt.len = NOP_IPV4_OPT;
+	opt.port = 0x0101;
+	opt.addr = 0x01010100;
```
Could this be just a `memset(&opt, IPOPT_NOOP, length);` + inserting the `IPOPT_END` as needed?

sypakine replied:
+1, Done.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2078,6 +2078,47 @@ nodeport_extract_dsr_v4(struct __ctx_buff *ctx,
+		return DROP_INVALID;
+	if (opt.type != DSR_IPV4_OPT_TYPE)
+		return CTX_ACT_OK;
+
```
What if we do the NOPing in `nodeport_extract_dsr_v4()` instead, where we already identified that the packet contains our option?

sypakine replied:
Done.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1998,9 +1998,46 @@ static __always_inline int encap_geneve_dsr_opt4(struct __ctx_buff *ctx, int l3_
+	if (ctx_load_bytes(ctx, ETH_HLEN + sizeof(struct iphdr), &opt, sizeof(opt)) < 0)
+		return DROP_INVALID;
+	if (opt.type != DSR_IPV4_OPT_TYPE)
+		return CTX_ACT_OK;
```
You could just have `nodeport_extract_dsr_v4()` pass in the options instead, and remove these parts.

sypakine replied:
Done.

julianwiedmann commented on bpf/tests/tc_nodeport_lb4_dsr_backend_remove_ip_option.c:
```diff
@@ -0,0 +1,440 @@
+	if (opt->port != 0x0101)
+		test_fatal("port in DSR IP option has changed: got type %d, want %d", opt->port, 0x0101);
+	if (opt->addr != 0x01010100)
+		test_fatal("addr in DSR IP option has change: got type %d, want %d", opt->addr, 0x01010100);
```
Could all this be just a `memcmp()` ?

sypakine replied:
Done.

julianwiedmann reviewed (commented):
Just two smaller remarks, otherwise this looks good on a quick glance.

sypakine committed:
Remove IP Option before forwarding the traffic to the backend endpoint [Optional]

Before forwarding the traffic to the endpoint
1. pad the DSR IP option with NOP options
2. recalculate the L3 checksum

Fixes: #29900

This change is a fix for backends that drop packets with IP options.

---

Release note:

Minor Changes:
- Remove IP Option before forwarding the traffic to the backend endpoint [Optional]

Signed-off-by: markstjohn@google.com

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2067,6 +2092,13 @@ nodeport_extract_dsr_v4(struct __ctx_buff *ctx,
 			*addr = bpf_ntohl(opt.addr);
 			*port = bpf_ntohs(opt.port);
+
+#ifdef REMOVE_DSR_IP_OPTION
+			int ret = remove_dsr_ip_opt_v4(ctx, ip4, &opt);
+			if (IS_ERR(ret))
+				return ret;
+#endif /* REMOVE_DSR_IP_OPTION */
```
Think we want a bit of compile coverage for `REMOVE_DSR_IP_OPTION` in bpf/Makefile. Adding it to some complexity test configs for `bpf_host` / `bpf_xdp` would be cool too.

julianwiedmann reviewed (commented):
Thanks - had a look at the latest diff, looks good! One testing remark inline.

There's some linter complaints in CI, think they should be easy to address. Might need to open-code the define for `IPOPT_NOOP` in the test, seems as if it doesn't like the `ip.h` include.
